cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/DefaultArtifactManager.java

```diff
@@ -0,0 +1,148 @@
+    // A URLClassLoader with no URLs and with a null parent will load class from bootstrap ClassLoader only.
+    this.bootstrapClassLoader = new URLClassLoader(new URL[0], null);
+  }
+
+  public List<ArtifactInfo> listArtifacts(NamespaceId namespaceId) throws IOException {
+    List<ArtifactDetail> artifactDetails = artifactStore.getArtifacts(namespaceId);
+    return Lists.transform(artifactDetails, new Function<ArtifactDetail, ArtifactInfo>() {
+      @Nullable
+      @Override
+      public ArtifactInfo apply(@Nullable ArtifactDetail input) {
+        // transform artifactDetail to artifactInfo
+        return new ArtifactInfo(input.getDescriptor().getArtifactId(),
+                                input.getMeta().getClasses(), input.getMeta().getProperties());
+      }
+    });
+  }
+
+  /**
+   * Ensures that the logged-in user has a {@link Action privilege} on the specified dataset instance.
+   *
+   * @param artifactId the {@link co.cask.cdap.proto.id.ArtifactId} to check for privileges
+   * @throws UnauthorizedException if the logged in user has no {@link Action privileges} on the specified dataset
+   */
+  private void ensureAccess(co.cask.cdap.proto.id.ArtifactId artifactId) throws Exception {
+    // No authorization for system artifacts
+    if (NamespaceId.SYSTEM.equals(artifactId.getParent())) {
+      return;
+    }
+    Principal principal = authenticationContext.getPrincipal();
+    Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
+    if (!filter.apply(artifactId)) {
+      throw new UnauthorizedException(principal, artifactId);
+    }
+  }
+
+
+  /**
+   * Create a class loader with artifact jar unpacked contents and parent for this classloader is the supplied
+   * parentClassLoader, if that parent classloader is null, bootstrap classloader is used as parent.
+   * This is a closeabled classloader, caller should call close when he is done using it, during close directory
+   * cleanup will be performed.
+   * @param namespace
+   * @param artifactInfo
+   * @param parentClassLoader
+   * @return CloseableClassLoader
+   * @throws Exception if artifact is not found or there were any error while getting artifact or if its unauthorized
+   */
+  public CloseableClassLoader createClassLoader(
+    NamespaceId namespace, ArtifactInfo artifactInfo, @Nullable ClassLoader parentClassLoader) throws Exception {
+
+    ensureAccess(new ArtifactId(namespace.getNamespace(), artifactInfo.getName(), artifactInfo.getVersion()));
+    ArtifactDetail artifactDetail = artifactStore.getArtifact(
+      new co.cask.cdap.proto.id.ArtifactId(namespace.getNamespace(),
+                                           artifactInfo.getName(), artifactInfo.getVersion()).toId());
+
+    final File unpackedDir = DirUtils.createTempDir(tmpDir);
+    BundleJarUtil.unJar(artifactDetail.getDescriptor().getLocation(), unpackedDir);
```

Comment:
So this method could potentially take a long time, depending on the size and number of files in the artifact jar. Probably better to document it as if someone call this method inside a short transaction, this can ended up in timeout frequently.